### PR TITLE
Document libopenblas deadlock issues and workaround

### LINKED ISSUE #1788: Deadlock in tgen (liblapack)
On my local machine, tgen is deadlocking at startup in shadow. It starts several threads, which appear to all be in a loop calling `sched_yield` and `rdtsc`; maybe some kind of spin lock? Here's a stack trace from one of the `sched_yield` calls:

```
(gdb) bt
<snip>
#14 0x00007ffff7fb8d88 in sched_yield (a=140737336181520, b=0, c=0, d=0, e=0, f=140737104509472)
    at /home/jnewsome/projects/shadow/dev/src/lib/libc_preload/syscall_wrappers.c:786
#15 0x00007ffff56f0286 in ?? () from /lib/x86_64-linux-gnu/liblapack.so.3
#16 0x00007ffff7458609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#17 0x00007ffff79ac293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Experimentally, changing `sched_yield` to move time forward eventually lets tgen move forward, but it seems strange that tgen has this spin-wait. It appears to be at the start of some `liblapack` worker thread, but I'm not familiar with `liblapack`, and I don't see where in tgen it's getting invoked.

If I attach gdb to tgen when it starts, and set a breakpoint on `clone`, it looks like the thread is created through some global initializer:

```
(gdb) bt
#0  clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:50
#1  0x00007ffff74602ec in create_thread (pd=pd@entry=0x7fffe9218700, attr=attr@entry=0x7fffffffea60, stopped_start=stopped_start@entry=0x7fffffffea5e, stackaddr=stackaddr@entry=0x7fffe91f9b40, thread_ran=thread_ran@entry=0x7fffffffea5f)
    at ../sysdeps/unix/sysv/linux/createthread.c:101
#2  0x00007ffff7461e10 in __pthread_create_2_1 (newthread=<optimized out>, attr=<optimized out>, start_routine=<optimized out>, arg=<optimized out>) at pthread_create.c:817
#3  0x00007ffff56f959c in blas_thread_init () from /lib/x86_64-linux-gnu/liblapack.so.3
#4  0x00007ffff4fcd083 in gotoblas_init () from /lib/x86_64-linux-gnu/liblapack.so.3
#5  0x00007ffff7fe0b8a in ?? () from /lib64/ld-linux-x86-64.so.2
#6  0x00007ffff7fe0c91 in ?? () from /lib64/ld-linux-x86-64.so.2
#7  0x00007ffff7fd013a in ?? () from /lib64/ld-linux-x86-64.so.2
#8  0x0000000000000002 in ?? ()
#9  0x00007fffffffed92 in ?? ()
#10 0x00007fffffffedb3 in ?? ()
#11 0x0000000000000000 in ?? ()
```

I suppose the next step is to look at the source of liblapack to find the source of the thread's starting function, which should have the wait-loop.

Btw I'm not sure why this is only happening on my machine and not in the CI. I'm running Ubuntu 20.04, same as the tor test in the CI:

```
$ lsb_release --all
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.3 LTS
Release:	20.04
Codename:	focal
```

### PR DESCRIPTION
**Pull request by sporksmith**
Fixes #1788

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ can fix it.
        * [Disabling Side-channel Mitigations](sidechannels.md)
        * [Parallel Simulations](parallel_sims.md)
      * [Migrating Simulations from Shadow 1.x](migrating_from_1x.md)
+     * Application and Library-Specific Notes
+       * [libopenblas](libopenblas.md)
```
Comment (robgjansen) [inline docs/README.md:38]: The indent spacing between the "Performance Tuning" subsection and the "Developer Guides" section appears inconsistent with the rest of the outline.

File: docs/SUMMARY.md [hunk 2]
```diff
@@ -29,6 +29,8 @@
     - [Performance Tuning]()
         - [Disabling Side-channel Mitigations](sidechannels.md)
         - [Parallel Simulations](parallel_sims.md)
+    - [Application and Library-Specific Notes]()
+        - [libopenblas](libopenblas.md)
```
Comment (robgjansen) [inline docs/SUMMARY.md:33]: Indent spacing consistency.

File: docs/libopenblas.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+libopenblas, when compiled with pthread support, makes extensive use of
+spin-waiting in `sched_yield`-loops, which currently result in deadlock under
+Shadow.
+
+There are several known workarounds:
+
+* Use a different implementation of libblas. e.g. on Ubuntu, there are several
+  alternative packages that can [provide
+  libblas](https://packages.ubuntu.com/hirsute/libblas.so.3).  In particular,
```
Comment (robgjansen) [inline docs/libopenblas.md:17]: Does the linebreak here break the hyperlink in the rendered output?

File: docs/libopenblas.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+* Install libopenblas compiled without pthread support. e.g. on Ubuntu this can
+  be obtained by installing
+  [libopenblas0-serial](https://packages.ubuntu.com/hirsute/libopenblas0-serial)
+  instead of
+  [libopenblas0-pthread](https://packages.ubuntu.com/hirsute/libopenblas0-pthread).
+
+* Configure libopenblas to not use threads at runtime. This can be done by
+  setting the environment variable `OPENBLAS_NUM_THREADS=1`, in the process's
+  [environment](https://shadow.github.io/docs/guide/shadow_config_spec.html#hostshostnameprocessesenvironment)
+  attribute in the Shadow config.
```
Comment (robgjansen) [inline docs/libopenblas.md:29]: Would it be worth linking to `src/test/tor/minimal/tor-minimal.yaml` to show a practical example of setting the env?

**Review by robgjansen — APPROVED** [review]

File: docs/README.md (hunk 1 above)
Comment (sporksmith) [reply docs/README.md:38]: I *think* it's a github rendering bug. It looks fine locally and here: https://raw.githubusercontent.com/shadow/shadow/aee287fac7ed58b96e3015c879b8a87f14595315/docs/README.md

File: docs/SUMMARY.md (hunk 2 above)
Comment (sporksmith) [reply docs/SUMMARY.md:33]: github rendering bug. https://raw.githubusercontent.com/shadow/shadow/aee287fac7ed58b96e3015c879b8a87f14595315/docs/SUMMARY.md

File: docs/libopenblas.md (hunk 3 above)
Comment (sporksmith) [reply docs/libopenblas.md:17]: Nope.

PS ... -> View file to see rendered output)
PPS Is there a name for a ... menu? Pacman menu...?

File: docs/libopenblas.md (hunk 4 above)
Comment (sporksmith) [reply docs/libopenblas.md]: Done

File: docs/libopenblas.md (hunk 3 above)
Comment (robgjansen) [reply docs/libopenblas.md:17]: I don't think I've seen it called a Pacman menu, but I like it! :)

**Pull request merged**